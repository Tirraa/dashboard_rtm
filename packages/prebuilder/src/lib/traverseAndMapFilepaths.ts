import type { Stats } from 'fs';

import type { Filename, File, Path } from '../types/Metadatas';
import type { Thunks, Thunk } from '../types/LazyEvaluation';

type ArborescenceFragmentPromise = Promise<File[] | File>;

// https://github.com/vitest-dev/vitest/discussions/2484
const path = require('path');
const fs = require('fs/promises');

async function makeThunks(currentFolder: Path, __currentDeepPath: Path = currentFolder): Promise<Thunks<ArborescenceFragmentPromise>> {
  const files: Filename[] = await fs.readdir(currentFolder);
  const thunks: Thunks<ArborescenceFragmentPromise> = files.map((currentFilename: Filename) => {
    const maybeFilepath: Path = path.join(currentFolder, currentFilename);
    const thunk: Thunk<ArborescenceFragmentPromise> = async () => {
      const filepathStats: Stats = await fs.stat(maybeFilepath);
      return filepathStats.isDirectory()
        ? execute(await makeThunks(maybeFilepath, path.join(__currentDeepPath, currentFilename)))
        : ({ directory: __currentDeepPath, name: currentFilename } satisfies File);
    };
    return thunk;
  });

  return thunks;
}

const execute = async (thunks: Thunks<ArborescenceFragmentPromise>): Promise<File[]> => (await Promise.all(thunks.map((thunk) => thunk()))).flat();

const traverseAndMapFilepaths = async (rootFolder: Path): Promise<File[]> => execute(await makeThunks(rootFolder));

export default traverseAndMapFilepaths;
