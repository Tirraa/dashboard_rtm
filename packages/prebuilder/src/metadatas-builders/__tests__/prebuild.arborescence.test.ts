// eslint-disable-next-line import/no-extraneous-dependencies
import { describe, expect, it } from 'vitest';

import traverseAndMapFilepaths from '../../lib/traverseAndMapFilepaths';
import buildArborescence from '../arborescence';

const VALID_ARBORESCENCE_FOLDER_PATH = './packages/prebuilder/src/metadatas-builders/__tests__/fake_arborescence';

describe('buildArborescence', () => {
  it('should match snapshot', async () => {
    const arborescenceMap = await traverseAndMapFilepaths(VALID_ARBORESCENCE_FOLDER_PATH);
    const arborescence = buildArborescence(arborescenceMap, VALID_ARBORESCENCE_FOLDER_PATH);
    expect(arborescence).toMatchSnapshot();
  });

  it('should match snapshot', async () => {
    const arborescenceMap = await traverseAndMapFilepaths(VALID_ARBORESCENCE_FOLDER_PATH);
    const arborescence = buildArborescence(arborescenceMap);
    expect(arborescence).toMatchSnapshot();
  });
});
