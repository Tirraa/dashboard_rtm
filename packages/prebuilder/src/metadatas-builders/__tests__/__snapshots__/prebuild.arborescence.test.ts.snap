// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`buildArborescence > should match snapshot 1`] = `
[
  {
    "directoriesChain": [
      "fake-category",
      "en",
    ],
    "filename": "ignore-me.FAKE_EXT",
  },
  {
    "directoriesChain": [
      "fake-category",
      "en",
    ],
    "filename": "lp-00.mdx",
  },
  {
    "directoriesChain": [
      "fake-category",
    ],
    "filename": "ignore-me.FAKE_EXT",
  },
  {
    "directoriesChain": [
      "fake-category",
    ],
    "filename": "lp-00.mdx",
  },
  {
    "directoriesChain": [],
    "filename": "not-a-directory.FAKE_EXT",
  },
]
`;

exports[`buildArborescence > should match snapshot 2`] = `
[
  {
    "directoriesChain": [
      "packages",
      "prebuilder",
      "src",
      "metadatas-builders",
      "__tests__",
      "fake_arborescence",
      "fake-category",
      "en",
    ],
    "filename": "ignore-me.FAKE_EXT",
  },
  {
    "directoriesChain": [
      "packages",
      "prebuilder",
      "src",
      "metadatas-builders",
      "__tests__",
      "fake_arborescence",
      "fake-category",
      "en",
    ],
    "filename": "lp-00.mdx",
  },
  {
    "directoriesChain": [
      "packages",
      "prebuilder",
      "src",
      "metadatas-builders",
      "__tests__",
      "fake_arborescence",
      "fake-category",
    ],
    "filename": "ignore-me.FAKE_EXT",
  },
  {
    "directoriesChain": [
      "packages",
      "prebuilder",
      "src",
      "metadatas-builders",
      "__tests__",
      "fake_arborescence",
      "fake-category",
    ],
    "filename": "lp-00.mdx",
  },
  {
    "directoriesChain": [
      "packages",
      "prebuilder",
      "src",
      "metadatas-builders",
      "__tests__",
      "fake_arborescence",
    ],
    "filename": "not-a-directory.FAKE_EXT",
  },
]
`;

exports[`buildArborescence should match snapshot 1`] = `
[
  {
    "directoriesChain": [
      "fake-category",
      "en",
    ],
    "filename": "ignore-me.FAKE_EXT",
  },
  {
    "directoriesChain": [
      "fake-category",
      "en",
    ],
    "filename": "lp-00.mdx",
  },
  {
    "directoriesChain": [
      "fake-category",
    ],
    "filename": "ignore-me.FAKE_EXT",
  },
  {
    "directoriesChain": [
      "fake-category",
    ],
    "filename": "lp-00.mdx",
  },
  {
    "directoriesChain": [],
    "filename": "not-a-directory.FAKE_EXT",
  },
]
`;

exports[`buildArborescence should match snapshot 2`] = `
[
  {
    "directoriesChain": [
      "packages",
      "prebuilder",
      "src",
      "metadatas-builders",
      "__tests__",
      "fake_arborescence",
      "fake-category",
      "en",
    ],
    "filename": "ignore-me.FAKE_EXT",
  },
  {
    "directoriesChain": [
      "packages",
      "prebuilder",
      "src",
      "metadatas-builders",
      "__tests__",
      "fake_arborescence",
      "fake-category",
      "en",
    ],
    "filename": "lp-00.mdx",
  },
  {
    "directoriesChain": [
      "packages",
      "prebuilder",
      "src",
      "metadatas-builders",
      "__tests__",
      "fake_arborescence",
      "fake-category",
    ],
    "filename": "ignore-me.FAKE_EXT",
  },
  {
    "directoriesChain": [
      "packages",
      "prebuilder",
      "src",
      "metadatas-builders",
      "__tests__",
      "fake_arborescence",
      "fake-category",
    ],
    "filename": "lp-00.mdx",
  },
  {
    "directoriesChain": [
      "packages",
      "prebuilder",
      "src",
      "metadatas-builders",
      "__tests__",
      "fake_arborescence",
    ],
    "filename": "not-a-directory.FAKE_EXT",
  },
]
`;
